### PR TITLE
feat(whiteboard): support multi window

### PR DESCRIPTION
**Pull request by Cheerego7**

### CODE REVIEW

File: desktop/renderer-app/src/components/Whiteboard.less [hunk 1]
```diff
@@ -30,47 +30,20 @@
...
     bottom: 8px;
     left: 8px;
-}
 
-.zoom-controller-box {
-    position: absolute;
-    left: 76px;
-    z-index: 3;
-    bottom: 8px;
+    &.is-disabled {
```
Comment (BlackHole1) [inline desktop/renderer-app/src/components/Whiteboard.less:34]:
```suggestion
    &.invisible {
```

File: desktop/renderer-app/src/components/Whiteboard.less [hunk 2]
```diff
@@ -30,47 +30,20 @@
...
-    height: 24px;
-    display: flex;
-    justify-content: center;
-    align-items: center;
-    cursor: pointer;
-    background-color: white;
-    border-radius: 2px;
-    &:hover {
-        background: rgba(33, 35, 36, 0.1);
+    &.is-disabled {
```
Comment (BlackHole1) [inline desktop/renderer-app/src/components/Whiteboard.less:45]:
```suggestion
    &.invisible {
```

**Review by crimx — APPROVED** [review]

File: desktop/renderer-app/src/components/Whiteboard.tsx [hunk 3]
```diff
@@ -66,37 +73,31 @@ export const Whiteboard = observer<WhiteboardProps>(function Whiteboard({ whiteb
...
-                    <ZoomController room={room} />
-                </div>
                 <div className="whiteboard-writable-area">
                     <div className="tool-box-out">
                         <ToolBox room={room} />
                     </div>
-                    <div className="redo-undo-box">
+                    <div
+                        className={classNames("redo-undo-box", {
+                            "is-disabled": whiteboardStore.isWindowMaximization,
```
Comment (BlackHole1) [inline desktop/renderer-app/src/components/Whiteboard.tsx:82]:
```suggestion
                            "invisible": whiteboardStore.isWindowMaximization,
```

File: desktop/renderer-app/src/components/Whiteboard.tsx [hunk 4]
```diff
@@ -66,37 +73,31 @@ export const Whiteboard = observer<WhiteboardProps>(function Whiteboard({ whiteb
...
-                            <div
-                                className="page-preview-cell"
-                                onClick={whiteboardStore.showPreviewPanel}
-                            >
-                                <img src={pagesSVG} alt={"pages"} />
-                            </div>
-                        </div>
+                    <div
+                        className={classNames("page-controller-box", {
+                            "is-disabled": whiteboardStore.isWindowMaximization,
```
Comment (BlackHole1) [inline desktop/renderer-app/src/components/Whiteboard.tsx:89]:
```suggestion
                            "invisible": whiteboardStore.isWindowMaximization,
```

**Review by crimx — CHANGES_REQUESTED** [review]
Please make it pass the spell check CI

File: desktop/renderer-app/src/components/Whiteboard.tsx [hunk 5]
```diff
@@ -66,37 +73,31 @@ export const Whiteboard = observer<WhiteboardProps>(function Whiteboard({ whiteb
...
+                        })}
+                    >
+                        <ScenesController
+                            addScene={whiteboardStore.addScene}
+                            preScene={whiteboardStore.preScene}
+                            nextScene={whiteboardStore.nextScene}
+                            currentSceneIndex={whiteboardStore.currentSceneIndex}
+                            scenesCount={whiteboardStore.scenesCount}
+                            disabled={whiteboardStore.isFocusWindow}
```
Comment (BlackHole1) [inline desktop/renderer-app/src/components/Whiteboard.tsx]:
```suggestion
                            whiteboardStore={whiteboardStore}
```

Comment (BlackHole1) [reply desktop/renderer-app/src/components/Whiteboard.tsx]: @crimx Will this be a problem?

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 6]
```diff
@@ -87,10 +107,100 @@ export class WhiteboardStore {
         this.isShowPreviewPanel = show;
     };
 
+    public setFocusWindowManager = (isFocus: boolean): void => {
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts]:
```suggestion
    public updateFocusWindowManager = (isFocus: boolean): void => {
```

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 7]
```diff
@@ -87,10 +107,100 @@ export class WhiteboardStore {
...
+        this.isFocusWindow = isFocus;
+    };
+
+    public setSwitchMainView = async (): Promise<void> => {
+        if (this.windowManager && this.isFocusWindow) {
+            await this.windowManager.switchMainViewToWriter();
+        }
+    };
+
+    public setWindowMaximization = (isMaximization: boolean): void => {
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts]:
```suggestion
    public updateWindowMaximization = (isMaximization: boolean): void => {
```

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 8]
```diff
@@ -87,10 +107,100 @@ export class WhiteboardStore {
         this.isShowPreviewPanel = show;
     };
 
+    public setFocusWindowManager = (isFocus: boolean): void => {
+        this.isFocusWindow = isFocus;
+    };
+
+    public setSwitchMainView = async (): Promise<void> => {
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts]:
```suggestion
    public switchMainViewToWriter = async (): Promise<void> => {
```

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 9]
```diff
@@ -87,10 +107,100 @@ export class WhiteboardStore {
...
+        if (this.windowManager && this.isFocusWindow) {
+            await this.windowManager.switchMainViewToWriter();
+        }
+    };
+
+    public setWindowMaximization = (isMaximization: boolean): void => {
+        this.isWindowMaximization = isMaximization;
+    };
+
+    public addScene = (): void => {
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts]:
```suggestion
    public addSceneToWindowManager = (): void => {
```

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 10]
```diff
@@ -87,10 +107,100 @@ export class WhiteboardStore {
...
+            const currentScene = this.currentSceneIndex + 1;
+            const scenePath = this.room.state.sceneState.scenePath;
+            const pathName = this.scenesPathName(scenePath);
+
+            this.room.putScenes(pathName, [{}], currentScene);
+            this.windowManager.setMainViewSceneIndex(this.currentSceneIndex + 1);
+        }
+    };
+
+    public preScene = (): void => {
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts]:
```suggestion
    public previousSceneByWindowManager = (): void => {
```

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 11]
```diff
@@ -87,10 +107,100 @@ export class WhiteboardStore {
...
+        if (this.windowManager && this.currentSceneIndex > 0) {
+            this.windowManager.setMainViewSceneIndex(this.currentSceneIndex - 1);
+        }
+    };
+
+    public nextScene = (): void => {
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts]:
```suggestion
    public nextSceneByWindowManager = (): void => {
```

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 12]
```diff
@@ -87,10 +107,100 @@ export class WhiteboardStore {
...
+        if (this.windowManager && this.currentSceneIndex < this.scenesCount - 1) {
+            this.windowManager.setMainViewSceneIndex(this.currentSceneIndex + 1);
+        }
+    };
+
     private preloadPPTResource = debounce(async (pptSrc: string): Promise<void> => {
         await getCoursewarePreloader().preload(pptSrc);
     }, 2000);
 
+    public openDocsFile = async (
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts]:
```suggestion
    public openDocsFileInWindowManager = async (
```

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 13]
```diff
@@ -87,10 +107,100 @@ export class WhiteboardStore {
...
+        if (this.windowManager) {
+            await this.windowManager.addApp({
+                kind: BuildinApps.DocsViewer,
+                options: {
+                    scenePath,
+                    title,
+                    scenes: scenes,
+                },
+            });
+        }
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts]:
```suggestion
        await this.windowManager?.addApp({
            kind: BuildinApps.DocsViewer,
            options: {
                scenePath,
                title,
                scenes: scenes,
            },
        });
```

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 14]
```diff
@@ -87,10 +107,100 @@ export class WhiteboardStore {
...
+    };
+
+    public openMediaFile = async (resourceSrc: string, title: string): Promise<void> => {
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts]:
```suggestion
    public openMediaFileInWindowManager = async (resourceSrc: string, title: string): Promise<void> => {
```

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 15]
```diff
@@ -87,10 +107,100 @@ export class WhiteboardStore {
...
+            await this.windowManager.addApp({
+                kind: BuildinApps.MediaPlayer,
+                options: {
+                    title,
+                },
+                attributes: {
+                    src: resourceSrc,
+                },
+            });
+        }
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts]:
```suggestion
        await this.windowManager?.addApp({
            kind: BuildinApps.MediaPlayer,
            options: {
                title,
            },
            attributes: {
                src: resourceSrc,
            },
        });
```

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 16]
```diff
@@ -87,10 +107,100 @@ export class WhiteboardStore {
...
+        if (this.windowManager) {
+            this.windowManager.emitter.on("mainViewModeChange", mode => {
+                const isWindow = mode !== ViewVisionMode.Writable;
+                this.setFocusWindowManager(isWindow);
+                if (!isWindow && this.room) {
+                    this.updateCurrentSceneIndex(this.room.state.sceneState.index);
+                    this.updateScenesCount(this.room.state.sceneState.scenes.length);
+                }
+            });
+        }
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts]:
```suggestion
        this.windowManager?.emitter.on("mainViewModeChange", mode => {
            const isWindow = mode !== ViewVisionMode.Writable;
            this.setFocusWindowManager(isWindow);
            if (!isWindow && this.room) {
                this.updateCurrentSceneIndex(this.room.state.sceneState.index);
                this.updateScenesCount(this.room.state.sceneState.scenes.length);
            }
        });
```

Comment (BlackHole1) [reply desktop/renderer-app/src/stores/WhiteboardStore.ts]: miss off?

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 17]
```diff
@@ -87,10 +107,100 @@ export class WhiteboardStore {
...
+    };
+
+    public onBoxStateChange = (): void => {
+        if (this.windowManager) {
+            this.windowManager.emitter.on("boxStateChange", mode => {
+                const isMaximization = mode === "maximized";
+                this.setWindowMaximization(isMaximization);
+            });
+        }
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts]:
```suggestion
        this.windowManager?.emitter.on("boxStateChange", mode => {
            const isMaximization = mode === "maximized";
            this.setWindowMaximization(isMaximization);
        });
```

Comment (BlackHole1) [reply desktop/renderer-app/src/stores/WhiteboardStore.ts]: miss off?

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 18]
```diff
@@ -87,10 +107,100 @@ export class WhiteboardStore {
...
+                this.setFocusWindowManager(isWindow);
+                if (!isWindow && this.room) {
+                    this.updateCurrentSceneIndex(this.room.state.sceneState.index);
+                    this.updateScenesCount(this.room.state.sceneState.scenes.length);
+                }
+            });
+        }
+    };
+
+    public onBoxStateChange = (): void => {
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts]:
```suggestion
    public onWindowManagerBoxStateChange = (): void => {
```

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 19]
```diff
@@ -175,6 +287,15 @@ export class WhiteboardStore {
                     } catch (err) {
                         console.log(err);
                     }
+
+                    if (
+                        this.room &&
+                        this.windowManager &&
+                        this.windowManager.mainView.mode === ViewVisionMode.Writable
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts]:
```suggestion
                        this.windowManager?.mainView.mode === ViewVisionMode.Writable
```

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 20]
```diff
@@ -222,19 +343,41 @@ export class WhiteboardStore {
...
         }
     }
 
     public destroy(): void {
-        if (this.room) {
+        if (this.room && this.windowManager) {
             this.preloadPPTResource.cancel();
+            this.windowManager.destroy();
             this.room.callbacks.off();
         }
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts]:
```suggestion
        this.preloadPPTResource.cancel();
        this.windowManager?.destroy();
        this.room?.callbacks.off();
```

File: desktop/renderer-app/src/stores/WhiteboardStore.ts [hunk 21]
```diff
@@ -222,19 +343,41 @@ export class WhiteboardStore {
...
     }
+
+    private scenesPathName = (scenePath: string): string => {
+        const cells = scenePath.split("/");
+        const popCell = cells.pop();
+        if (popCell === "") {
+            cells.pop();
+        }
+        return cells.join("/");
+    };
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/WhiteboardStore.ts:374]: What does this mean? Can you give me an example of what the value of `scenePath` looks like?

File: web/flat-web/src/components/Whiteboard.less [hunk 22]
```diff
@@ -29,47 +29,20 @@
...
     bottom: 8px;
     left: 8px;
-}
 
-.zoom-controller-box {
-    position: absolute;
-    left: 76px;
-    z-index: 3;
-    bottom: 8px;
+    &.is-disabled {
```
Comment (BlackHole1) [inline web/flat-web/src/components/Whiteboard.less:33]: Same problem as above

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: desktop/renderer-app/src/components/Whiteboard.tsx (hunk 5 above)
Comment (crimx) [reply desktop/renderer-app/src/components/Whiteboard.tsx]: @BlackHole1 The component is pure so it should not be aware of the Whiteboard store. This optimization is possible if adding an extra layer to handle MobX values (like how we did with the ClouldStorage component) but probably not worth the effort for such a small component.

File: desktop/renderer-app/src/components/Whiteboard.tsx (hunk 3 above)
Comment (crimx) [reply desktop/renderer-app/src/components/Whiteboard.tsx:82]: We use `is-` prefix to indicate state classnames.

**Pull request merged**